src/structures/AuditLog.ts

```diff
@@ -0,0 +1,347 @@
+
+interface MessageDeleteEntry extends BaseEntry {
+  actionType: "messageDelete";
+  extra: {
+    channelId: Snowflake;
+    count: string;
+  };
+}
+
+interface MessageBulkDeleteEntry extends BaseEntry {
+  actionType: "messageBulkDelete";
+  extra: {
+    count: string;
+  };
+}
+
+interface MessagePinEntry extends BaseEntry {
+  actionType: "messagePin" | "messageUnpin";
+  extra: {
+    channelId: Snowflake;
+    messageId: Snowflake;
+  };
+}
+
+type Entry =
+  | MemberPruneEntry
+  | MemberMoveEntry
+  | MessagePinEntry
+  | MessageDeleteEntry
+  | MessageBulkDeleteEntry
+  | MemberDisconnectEntry
+  | ChannelOverwriteEntry
+  | NonExtraEntry;
+
+export interface ChangeKey {
+  name: string;
+  iconHash: string;
+  splashHash: string;
+  ownerId: Snowflake;
+  region: string;
+  afkChannelId: Snowflake;
+  afkTimeout: number;
+  mfaLevel: number;
+  verificationLevel: number;
+  explicitContentFilter: number;
+  defaultMessageNotifications: number;
+  vanityUrlCode: string;
+  roleAdd: Pick<role.Role, "id" | "name">[];
+  roleRemove: Pick<role.Role, "id" | "name">[];
+  pruneDeleteDays: number;
+  widgetEnabled: boolean;
+  widgetChannelId: Snowflake;
+  systemChannelId: Snowflake;
+  position: number;
+  topic: string;
+  bitrate: number;
+  permissionOverwrites: PermissionOverwrite[];
+  nsfw: boolean;
+  application_id: Snowflake;
+  rateLimitPerUser: number;
+  permissions: number;
+  permissionsNew: string;
+  color: number;
+  hoist: boolean;
+  mentionable: boolean;
+  allow: number;
+  allowNew: string;
+  deny: number;
+  denyNew: string;
+  code: string;
+  channelId: Snowflake;
+  inviterId: Snowflake;
+  maxUses: number;
+  uses: number;
+  maxAge: number;
+  temporary: boolean;
+  deaf: boolean;
+  mute: boolean;
+  nick: string;
+  avatarHash: string;
+  id: Snowflake;
+  type: number | string;
+  enableEmoticons: boolean;
+  expireBehavior: number;
+  expireGracePeriod: number;
+}
+
+export interface UnspecificChange<T extends keyof ChangeKey> {
+  newValue?: ChangeKey[T];
+  oldValue?: ChangeKey[T];
+  key: T;
+}
+
+type SpecificChange<T extends keyof ChangeKey> = T extends keyof ChangeKey
+  ? UnspecificChange<T>
+  : never;
+
+export type Change = SpecificChange<keyof ChangeKey>;
+
+export function parseAuditLog(
+  client: Client,
+  { audit_log_entries, integrations, users, webhooks }: auditLog.AuditLog,
+): AuditLog {
+  return {
+    users: users.map((user) => new User(client, user)),
+    entries: audit_log_entries.map((entry) => parseEntry(entry)),
+    integrations,
+    webhooks: webhooks.map((webhook) => parseWebhook(client, webhook)),
+  };
+}
+
+function parseEntry(entry: auditLog.Entry): Entry {
+  let extra;
+
+  if (entry.options) {
+    switch (entry.action_type) {
+      case 13:
+      case 14:
+      case 15:
+        extra = {
+          id: entry.options.id,
+          type: entry.options.type,
+          roleName: entry.options.role_name,
+        } as ChannelOverwriteMember | ChannelOverwriteRole;
```

Comment:
Unnecessary cast? This sort of casting should generally be avoided.

Reply:
Removed, at one point it was erroring wihtout.